### PR TITLE
Fix race in s3 multipart upload

### PR DESCRIPTION
**Pull request by vitlibar**
### Changelog category:
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry:
Fix race in s3 multipart upload. This race could cause the error `Part number must be an integer between 1 and 10000, inclusive. (S3_ERROR)` while restoring from a backup.

### CODE REVIEW

File: src/IO/WriteBufferFromS3.cpp [hunk 1]
```diff
@@ -122,7 +123,7 @@ void WriteBufferFromS3::nextImpl()
 
 void WriteBufferFromS3::allocateBuffer()
 {
-    if (total_parts_uploaded != 0 && total_parts_uploaded % request_settings.upload_part_size_multiply_parts_count_threshold == 0)
```
Comment (vitlibar) [inline src/IO/WriteBufferFromS3.cpp:125]: Here was the race. `total_parts_uploaded` was the number of completely uploaded parts, not just requested for upload. So when it was uploading a lot of data `total_parts_uploaded` and `upload_part_size` were updated with a delay which could lead to that error `part number is too big`.

**Review by antonio2368 — APPROVED** [review]
LGTM and nice find!

**Comment by vitlibar** [comment]
The test failures seem unrelated.

**Comment by orloffv** [comment]
this fix will be backported to 22.11?

**Pull request merged**